### PR TITLE
feat: Support selecting specific token indices when applying and training steering vector

### PR DESCRIPTION
**Pull request by Felhof**
This PR adds support for selecting specific token indices when applying a steering vector, and for training steering vectors at different token indices for different prompts.

### Steering at specific indices
 A `token_indices` parameter was added to the `apply` and `patch_activations` methods of `steering_vector`. This lets the user select specific token indices at which a steering vector should be applied. `token_indices` can be either: 
- a list or list slice of integers, containing the indices at which to apply steering.
- a tensor mask of 0s and 1s, where 1s denote the positions at which to apply steering.

The `min_token_index` parameter for `apply` and `patch_activations` is still supported. Passing `min_token_index=n` is equivalent to passing `token_indices=slice(n,None)`

**Example**
```
# the following will apply the steering vector at token indices 2, 4, and 7
target_token_indices = [2, 4, 7]
steering_vector.patch_activations(model, token_indices=target_token_indices)

# equivalently, token_indices can be a mask
target_token_indices = torch.tensor([1 if i in [2, 4, 7] else 0 for i in range(input_length)]
steering_vector.patch_activations(model, token_indices=target_token_indices)
```

### Training at different indices for different prompts
_Note: Updated on Feb 1st 2024_
There are two ways of passing custom indices to `train_steering_vector`.
- Via training data: `SteeringVectorTrainingSample` now optionally takes a `read_positive_token_index` and `read_negative_token_index`. The `training_data` argument of `train_steering_vector` can now also be a 4-tuple, with the last two values being the index of the positive and negative example respectively. 
- Via the token index argument: The `read_token_index` argument can now be a callable, mapping prompts to indices. It can still be an int, and by default will still be -1.

**Example using training data**
```
def get_x_index(prompt: str) -> int:
        tokenization = tokenizer.convert_ids_to_tokens(tokenizer.encode(prompt))
        return tokenization.index("ĠX")

training_data = [
        (
            "This is a short positive example. X <- probe here.",
            "This is a short negative example with different token length. X <- probe here.",
            get_x_index("This is a short positive example. X <- probe here."),
            get_x_index(
                "This is a short negative example with different token length. X <- probe here."
            ),
        ),
        (
            "Dummy text. This is a much longer positive example. X <- probe here. More dummy text.",
            "Dummy text. This is a much longer negative example with different token length. X <- probe here. More dummy text.",
            get_x_index(
                "Dummy text. This is a much longer positive example. X <- probe here. More dummy text."
            ),
            get_x_index(
                "Dummy text. This is a much longer negative example with different token length. X <- probe here. More dummy text."
            ),
        ),
    ]

steering_vector = train_steering_vector(
        model, tokenizer, training_data, layers=[2, 3, 4]
)
```

**Example using read_token_index data**
```
def get_x_index(prompt: str) -> int:
        tokenization = tokenizer.convert_ids_to_tokens(tokenizer.encode(prompt))
        return tokenization.index("ĠX")

training_data = [
        (
            "This is a short positive example. X <- probe here.",
            "This is a short negative example with different token length. X <- probe here.",
        ),
        (
            "Dummy text. This is a much longer positive example. X <- probe here. More dummy text.",
            "Dummy text. This is a much longer negative example with different token length. X <- probe here. More dummy text.",
        ),
    ]

# note how the get_x_index function is passed to train_steering_vector
steering_vector = train_steering_vector(
        model, tokenizer, training_data, layers=[2, 3, 4], read_token_index=get_x_index
)
```

### CODE REVIEW

File: steering_vectors/steering_vector.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 from contextlib import contextmanager
 from dataclasses import dataclass, replace
-from typing import Any, Callable, Generator, Optional, overload
+from typing import Any, Callable, Generator, List, Optional, overload
```
Comment (chanind) [inline steering_vectors/steering_vector.py:3]: nit: no need to use `List`, we can just use `list` since we're targeting Python 3.10+

File: steering_vectors/steering_vector.py [hunk 2]
```diff
@@ -101,6 +112,7 @@ def apply(
         operator: Optional[PatchOperator] = None,
         multiplier: float = 1.0,
         min_token_index: int = 0,
+        token_indices: List[int] | slice | Tensor | None = None,
```
Comment (chanind) [inline steering_vectors/steering_vector.py:115]: nit: also need to update the docstring with this new param so it will be reflected in API docs

File: steering_vectors/steering_vector.py [hunk 3]
```diff
@@ -45,7 +45,8 @@ def patch_activations(
         layer_config: Optional[ModelLayerConfig] = None,
         operator: Optional[PatchOperator] = None,
         multiplier: float = 1.0,
-        min_token_index: int = 0,
+        min_token_index: int | None = None,
+        token_indices: List[int] | slice | Tensor | None = None,
```
Comment (chanind) [inline steering_vectors/steering_vector.py:49]: Here as well, need to update the docstring below with this new option

File: steering_vectors/steering_vector.py [hunk 4]
```diff
@@ -68,6 +69,16 @@ def patch_activations(
...
         """
+        assert (min_token_index is None) or (
+            token_indices is None
+        ), "Can not pass both min_token_index and token_indices"
+        if isinstance(token_indices, Tensor):
+            assert torch.all(
+                torch.logical_or(token_indices == 0, token_indices == 1)
+            ), "token_indices tensor must be a mask (containing only 0s and 1s)"
+        token_indices = (
+            token_indices if token_indices is not None else slice(min_token_index, None)
```
Comment (chanind) [inline steering_vectors/steering_vector.py:80]: 👍 This is cool, didn't know you could do this with `slice(None, None)`!

File: steering_vectors/steering_vector.py [hunk 5]
```diff
@@ -181,8 +194,13 @@ def hook_fn(_m: Any, _inputs: Any, outputs: Any) -> Any:
...
         if operator is not None:
             delta = operator(original_tensor, act)
-        mask = torch.ones(original_tensor.shape[1])
-        mask[:min_token_index] = 0
+        if isinstance(token_indices, Tensor):
+            mask = token_indices
+        elif token_indices is not None:
+            mask = torch.zeros(original_tensor.shape[1])
+            mask[token_indices] = 1
+        else:
```
Comment (chanind) [inline steering_vectors/steering_vector.py:202]: It looks like we don't need to handle the `token_indices is None` case, since above it gets initialized to `slice(None, None)`, if nothing is passed in. I think the type definition for this function can remove the `| None = None` part too.

File: steering_vectors/train_steering_vector.py [hunk 6]
```diff
@@ -27,7 +27,7 @@ def train_steering_vector(
     layer_type: LayerType = "decoder_block",
     layer_config: Optional[ModelLayerConfig] = None,
     move_to_cpu: bool = False,
-    read_token_index: int = -1,
+    read_token_index: int | List[int] | Tensor = -1,
```
Comment (chanind) [inline steering_vectors/train_steering_vector.py:30]: nit: here as well, just use `list` since we're using 3.10+ (we should add a linting rule for this...)

File: steering_vectors/train_steering_vector.py [hunk 7]
```diff
@@ -56,9 +56,18 @@ def train_steering_vector(
     layer_config = guess_and_enhance_layer_config(model, layer_config, layer_type)
     pos_activations: dict[int, list[Tensor]] = defaultdict(list)
     neg_activations: dict[int, list[Tensor]] = defaultdict(list)
+    n_pairs = len(training_samples)
+    if not isinstance(read_token_index, int):
+        pos_token_indices = read_token_index[:n_pairs]
+        neg_token_indices = read_token_index[n_pairs:]
```
Comment (chanind) [inline steering_vectors/train_steering_vector.py:62]: I'm a bit confused by this code. I assumed that if a list is passed for `read_token_index`, that would either mean that either:
-  a training pair is generated for every item in the list (e.g. we want a sample for both indices `-1` and `-2`).
- We somehow aggregate the activations from each item in the list to still get a single pos/neg pair for each training sample (e.g. we would take the mean of activations at indices `-1` and `-2`)

It looks like this is instead assuming that `read_token_index` is 2x the length of the training examples, and is setting a specific index for every example, where the first 0-n are for the first 0-n positive prompts, and n-2*n is for the corresponding negative prompts?

IMO if we want to still extract a single index for each prompt, but want to have the ability to set a different index per training sample, we shouldn't do it via passing in a 2x length list to the `read_token_index` param, since I feel this is hard to communicate to the user and easy to accidentally mess up. I would instead recommend adding something like `read_positive_token_index` and `read_negative_token_index` to `SteeringVectorTrainingSample`, and have the user pass in a `SteeringVectorTrainingSample` instance instead of a pos/neg prompt tuple. This way, it's clear that these values apply to the specific training sample they're attached to, and doesn't require communicating to the user that they need this 2x list length format. This is an advanced use-case, and for the basic use-case users could still use the tuple of `(pos, neg)` instead. If a `read_positive_token_index` or `read_negative_token_index` is provided for the `SteeringVectorTrainingSample`, that would override the `read_token_index` passed to the function.

So, it would look something like:

```python
@dataclass
class SteeringVectorTrainingSample:
    positive_prompt: str
    negative_prompt: str
    read_positive_token_index: Optional[int] = None
    read_negative_token_index: Optional[int] = None
```

What do you think of this proposal? Apologies if I'm misunderstanding the intention of this code!

File: tests/test_steering_vector.py [hunk 8]
```diff
@@ -113,6 +116,57 @@ def test_SteeringVector_patch_activations_with_min_token_index(
...
+    [
+        ([2, 4, 7], [i for i in range(13) if i not in [2, 4, 7]]),
+        (
+            torch.tensor([1 if i in [2, 4, 7] else 0 for i in range(13)]),
+            [i for i in range(13) if i not in [2, 4, 7]],
+        ),
+    ],
+)
+@torch.no_grad()
+def test_SteeringVector_patch_activations_with_token_indices(
```
Comment (chanind) [inline tests/test_steering_vector.py:131]: 👍  Great test! It would also be good to test the other types of inputs this can take (slice and Tensor). You can just copy/paste this test and change `target_token_indices` IMO

File: steering_vectors/train_steering_vector.py (hunk 7 above)
Comment (chanind) [reply steering_vectors/train_steering_vector.py:62]: Another thought: Maybe the easiest way to achieve this is just to allow passing in a callable to `read_token_index` which takes in the prompt and returns an index. So, something like:

```python
read_token_index: int | Callable[[str], int] = -1
``` 

Then, the user can pass in a function to do whatever crazy logic they want, and it's pretty easy for us to implement and for the user to understand. For instance, if you wanted to find a specific token in the text, you could calculate that in the callable.

Comment (Felhof) [reply steering_vectors/train_steering_vector.py:62]: > What do you think of this proposal?

You understood correctly and I agree with the concern that my current implementation isn't very intuitive to understand. I will instead add a Callable and extend `SteeringVectorTrainingSample` like you first suggested

Comment (Felhof) [reply steering_vectors/train_steering_vector.py]: Done in https://github.com/steering-vectors/steering-vectors/pull/21/commits/76a9b9857d55a64d271626617708b33ad0b0e98e

File: steering_vectors/train_steering_vector.py (hunk 6 above)
Comment (Felhof) [reply steering_vectors/train_steering_vector.py]: Done in [83b0ccf](https://github.com/steering-vectors/steering-vectors/pull/21/commits/83b0ccff5fa9fdedafe08830592102e55fd0b980)

File: steering_vectors/steering_vector.py (hunk 3 above)
Comment (Felhof) [reply steering_vectors/steering_vector.py]: Done in [83b0ccf](https://github.com/steering-vectors/steering-vectors/pull/21/commits/83b0ccff5fa9fdedafe08830592102e55fd0b980)

File: steering_vectors/steering_vector.py (hunk 2 above)
Comment (Felhof) [reply steering_vectors/steering_vector.py]: Done in [83b0ccf](https://github.com/steering-vectors/steering-vectors/pull/21/commits/83b0ccff5fa9fdedafe08830592102e55fd0b980)

File: steering_vectors/steering_vector.py (hunk 1 above)
Comment (Felhof) [reply steering_vectors/steering_vector.py]: Done in [83b0ccf](https://github.com/steering-vectors/steering-vectors/pull/21/commits/83b0ccff5fa9fdedafe08830592102e55fd0b980)

File: tests/test_steering_vector.py (hunk 8 above)
Comment (Felhof) [reply tests/test_steering_vector.py:136]: I'm using pytest parametrization to do it all in one test now. Done in [9eb39c5](https://github.com/steering-vectors/steering-vectors/pull/21/commits/9eb39c5ed9b26e48d5e7bc05bf092befad50593c)

File: steering_vectors/steering_vector.py (hunk 5 above)
Comment (Felhof) [reply steering_vectors/steering_vector.py]: Done in [0343465](https://github.com/steering-vectors/steering-vectors/pull/21/commits/03434650e28b25533732909da51b682d03017a4c)

File: tests/test_steering_vector.py (hunk 8 above)
Comment (chanind) [reply tests/test_steering_vector.py:136]: nice! even cleaner

File: steering_vectors/steering_vector.py [hunk 9]
```diff
@@ -45,13 +45,16 @@ def patch_activations(
         layer_config: Optional[ModelLayerConfig] = None,
         operator: Optional[PatchOperator] = None,
         multiplier: float = 1.0,
-        min_token_index: int = 0,
+        min_token_index: int | None = None,
+        token_indices: list[int] | slice | Tensor | None = None,
```
Comment (chanind) [inline steering_vectors/steering_vector.py:49]: I was going to write here that it's better to use `Optional[types] = None` rather than `types | None = None` since it indicates the intention that it's an optional parameter, but after some searching it looks like I'm wrong on this and the best practice now is to replace all the `Optional[types]` with `types | None`. I added an issue to remove all the `Optional` stuff throughout the codebase here: https://github.com/steering-vectors/steering-vectors/issues/23

File: steering_vectors/train_steering_vector.py [hunk 10]
```diff
@@ -16,18 +16,22 @@
 class SteeringVectorTrainingSample(NamedTuple):
```
Comment (chanind) [inline steering_vectors/train_steering_vector.py:16]: nit: we should change this to a dataclass at this point since these optional params don't make a lot of sense as tuples IMO

File: steering_vectors/train_steering_vector.py [hunk 11]
```diff
@@ -56,18 +60,39 @@ def train_steering_vector(
     layer_config = guess_and_enhance_layer_config(model, layer_config, layer_type)
     pos_activations: dict[int, list[Tensor]] = defaultdict(list)
     neg_activations: dict[int, list[Tensor]] = defaultdict(list)
+    if len(training_samples[0]) == 2:
+        training_samples = [
+            SteeringVectorTrainingSample(sample[0], sample[1], None, None)
+            for sample in training_samples
+        ]
+
+    def get_token_index(
```
Comment (chanind) [inline steering_vectors/train_steering_vector.py:69]: nit: This internal function doesn't reference anything in the outer scope, so I'd recommend moving this into a top-level function called `_get_token_index()`. This will clean up the `train_steering_vector()` function, make it clearer that this function is a simple pure function of its inputs, and make it possible to test this function explicitly in unit tests.

File: steering_vectors/train_steering_vector.py (hunk 7 above)
Comment (chanind) [reply steering_vectors/train_steering_vector.py]: Nice, doing both also works!

File: steering_vectors/train_steering_vector.py (hunk 10 above)
Comment (Felhof) [reply steering_vectors/train_steering_vector.py]: Resolved by [be9fb91](https://github.com/steering-vectors/steering-vectors/pull/21/commits/be9fb919720b59a0d099cf0f1b4a6de0c6676ba7)

File: steering_vectors/train_steering_vector.py [hunk 12]
```diff
@@ -16,18 +16,22 @@
...
 
 
 @torch.no_grad()
 def train_steering_vector(
     model: nn.Module,
     tokenizer: PreTrainedTokenizerBase,
-    training_samples: list[SteeringVectorTrainingSample] | list[tuple[str, str]],
+    training_samples: list[SteeringVectorTrainingSample]
+    | list[tuple[str, str]]
+    | list[tuple[str, str, Optional[int], Optional[int]]],
```
[the remark this answers is not in the document]
Comment (Felhof) [reply steering_vectors/train_steering_vector.py]: Resolved by [be9fb91](https://github.com/steering-vectors/steering-vectors/pull/21/commits/be9fb919720b59a0d099cf0f1b4a6de0c6676ba7)

File: steering_vectors/train_steering_vector.py (hunk 11 above)
Comment (Felhof) [reply steering_vectors/train_steering_vector.py]: Resolved by [50aed1b](https://github.com/steering-vectors/steering-vectors/pull/21/commits/50aed1b17e6b6beb9081b78b72d3373bd3465fa2)

**Comment by chanind** [comment]
Awesome work @Felhof!
